handlers/handlers.go

```diff
@@ -34,28 +36,71 @@ func (h *Handler) MemesIndex(w http.ResponseWriter, r *http.Request) {
 }
 
 // MemesCreate creates a meme
+// TODO: function too long
 func (h *Handler) MemesCreate(w http.ResponseWriter, r *http.Request) {
-	meme := &models.Meme{URL: r.FormValue("url")}
+	meme := &models.Meme{OriginalUrl: r.FormValue("url")}
+	resp, err := http.Get(meme.OriginalUrl)
+	if err != nil {
+		log.Fatalf("%s %s", "http error", meme.OriginalUrl)
+		return
+	}
+	if resp.StatusCode != 200 || resp.ContentLength > maxImageContentLength {
+		log.Fatalf("%s %s", "error status or response too big", resp.StatusCode)
+		return
+	}
+	contentTypeValid := false
+	for _, contentType := range allowedContentTypes {
+		if contentType == resp.Header.Get("Content-Type") {
+			contentTypeValid = true
+			break
+		}
+	}
+	if !contentTypeValid {
+		log.Fatalf("invalid content type header")
+		return
+	}
+	if err := meme.SetData(resp.Body, resp.Header.Get("Content-Type")); err != nil {
+		log.Fatal(err)
+	}
```

Comment:
Could be. Another way is a Builder/Factory class that will be a mould for Memes - either from URL or bin data (`/upload`). We are still missing validations which won't be easy 🙉